src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -458,33 +464,46 @@ internal bool LocalFindAndPop(IThreadPoolWorkItem callback)
 
         public IThreadPoolWorkItem Dequeue(ThreadPoolWorkQueueThreadLocals tl, ref bool missedSteal)
         {
-            WorkStealingQueue localWsq = tl.workStealingQueue;
             IThreadPoolWorkItem callback;
-
-            if ((callback = localWsq.LocalPop()) == null && // first try the local queue
-                !workItems.TryDequeue(out callback)) // then try the global queue
+            int wsqActiveObserved = ThreadPoolGlobals.wsqActive;
+            if(wsqActiveObserved > 0)
             {
-                // finally try to steal from another thread's local queue
-                WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
-                int c = queues.Length;
-                Debug.Assert(c > 0, "There must at least be a queue for this thread.");
-                int maxIndex = c - 1;
-                int i = tl.random.Next(c);
-                while (c > 0)
+                WorkStealingQueue localWsq = tl.workStealingQueue;
+
+                if ((callback = localWsq.LocalPop()) == null && // first try the local queue
+                    !workItems.TryDequeue(out callback)) // then try the global queue
                 {
-                    i = (i < maxIndex) ? i + 1 : 0;
-                    WorkStealingQueue otherQueue = queues[i];
-                    if (otherQueue != localWsq && otherQueue.CanSteal)
+                    // finally try to steal from another thread's local queue
+                    WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
+                    int c = queues.Length;
+                    Debug.Assert(c > 0, "There must at least be a queue for this thread.");
+                    int maxIndex = c - 1;
+                    int i = tl.random.Next(c);
+                    while (c > 0)
                     {
-                        callback = otherQueue.TrySteal(ref missedSteal);
-                        if (callback != null)
+                        i = (i < maxIndex) ? i + 1 : 0;
+                        WorkStealingQueue otherQueue = queues[i];
+                        if (otherQueue != localWsq && otherQueue.CanSteal)
                         {
-                            break;
+                            callback = otherQueue.TrySteal(ref missedSteal);
+                            if (callback != null)
+                            {
+                                break;
+                            }
                         }
+                        c--;
+                    }
+                    if((callback == null) && !missedSteal)
+                    {
+                        Interlocked.CompareExchange(ref ThreadPoolGlobals.wsqActive, wsqActiveObserved - 1, wsqActiveObserved);
```

Comment:
The intended design is:
- The compare exchange only allows one thread to decrement ThreadPoolGlobals.wsqActive.  The value of ThreadPoolGlobals.wsqActive must not have changed since it was read before initiating the scan through WSQL for the decrement to succeed, if the value has changed the compare and exchange will fail.
- The value of wsqNowActive=2 is such that one observer might miss the new item.  A second should not.  I had originally thought this would need to be 3, but I do not see how that is required or would fix anything.

ThreadPoolGlobals.wsqActive needs to ordered in Enqueue().  It is required that the new work item be observable in the WSQ before reading ThreadPoolGlobals.wsqActive.  If they are out of order this would not be sufficient.  I am assuming there is a barrier operation included in workStealingQueue.LocalPush().  Perhaps it would be better to mark wsqActive volatile to prevent reordering of operations.

Ultimately I want reading ThreadPoolGlobals.wsqActive to be a lightweight operation.  If I can avoid globally marking it volatile that would be helpful.  Is there a way for Enqueue() to treat as volatile w/o Dequeue() treating as volatile?  Would 
```C#
ref volatile int wsqActiveVolatile = ThreadPoolGlobals.wsqActive; 
if(wsqActiveVolatile != ThreadPoolGlobals.wsqNowActive)
```
 work to guarantee ordering?